smarts/core/local_traffic_provider.py

```diff
@@ -0,0 +1,1231 @@
+                    break
+                self._in_front_after_cutin_secs += dt
+                if self._in_front_after_cutin_secs < self._cutin_hold_secs:
+                    break
+            self._cutting_into = None
+            self._in_front_secs = 0
+            if lw.agent_gap and self._should_cutin(lw):
+                best_lw = lw
+                self._cutting_into = lw
+            elif lw.adj_time_left > best_lw.adj_time_left or (
+                lw.adj_time_left == best_lw.adj_time_left
+                and (
+                    (lw.lane == self._dest_lane and self._offset < self._dest_offset)
+                    or (lw.ttre > best_lw.ttre and idx < best_lw.lane.index)
+                )
+            ):
+                best_lw = lw
+        if best_lw.lane != self._lane and not self._cutting_into:
+            self._cutting_into = best_lw
+        self._target_lane_win = best_lw
+
+    def _compute_lane_speeds(self):
+        def _get_radius(lane: RoadMap.Lane) -> float:
+            l_offset = self._owner._cached_lane_offset(self._state, lane)
+            # we round the offset in an attempt to reduce the unique hits on the LRU caches...
+            l_offset = round(l_offset)
+            l_width, _ = lane.width_at_offset(l_offset)
+            return lane.curvature_radius_at_offset(
+                l_offset, lookahead=max(math.ceil(2 * l_width), 2)
+            )
+
+        self._lane_speed = dict()
+        my_radius = _get_radius(self._lane)
+        for l in self._lane.road.lanes:
+            ratio = 1.0
+            if l != self._lane and abs(my_radius) < 1e5:
+                l_radius = _get_radius(l)
+                if abs(l_radius) < 1e5:
+                    ratio = l_radius / my_radius
+                    if ratio < 0:
+                        ratio = 1.0
+            self._lane_speed[l.index] = (ratio * self.speed, ratio * self.acceleration)
+
+    def _slow_for_curves(self):
+        # XXX:  this may be too expensive.  if so, we'll need to precompute curvy spots for routes
+        lookahead = math.ceil(1 + math.log(self._target_speed))
+        # we round the offset in an attempt to reduce the unique hits on the LRU caches...
+        rounded_offset = round(self._offset)
+        radius = self._lane.curvature_radius_at_offset(rounded_offset, lookahead)
+        # pi/2 radian right turn == 6 m/s with radius 10.5 m
+        # TODO:  also depends on vehicle type (traction, length, etc.)
+        self._target_speed = min(abs(radius) * 0.5714, self._target_speed)
+
+    def _check_speed(self):
+        target_lane = self._target_lane_win.lane
+        if target_lane.speed_limit is None:
+            self._target_speed = self.speed
+            self._slow_for_curves()
+            return
+        self._target_speed = target_lane.speed_limit
+        self._target_speed *= self._speed_factor
+        self._slow_for_curves()
+        max_speed = float(self._vtype.get("maxSpeed", 55.55))
```

Comment:
I suppose my final concern is that a lot of these magic numbers are related to a set of default undocumented/undeclared models.

Reply:
I added some comments about the magic numbers.